Add exhaustive testing to ValueRanges, fix bugs

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95063
* #94985
* #94944
* __->__ #94939
* #94906

Since I didn't want to deal with nondeterministic tests, I went the exhaustive testing route for a fixed list of constants to look at. The tests generate random ranges, propagate the range through the function, and then pick elements in the range and check that the result on the operation is in the resulting range. This caught bugs in log, sqrt and pow.

My resolution for pow was a little special, because I had trouble figuring out the correct semantics under all inputs domains. Instead, I picked two input domains (pow on two point ranges, and pow where exponent is known) and only implemented those. Everything else we give up. I think this is unlikely to affect perf.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

ezyang committed:
Add fuzz testing to ValueRanges, fix bugs

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

lezcano commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -261,31 +261,36 @@ def exp(x):
+                i = cls.mul(i, a)
+            return i
+        else:
+            # This is fairly difficult to analyze, so give up for anything
+            # complicated
```
Right, I see we were missing a couple cases here, yep.

lezcano replied:
For future reference, it's probably enough to analyse the signs of a and b separately and have those 4 cases, but yeah, probably not worth the effort.

ezyang replied:
PRs welcome! The problem I had was that you also have different regimes for a < 1 vs > 1 vs = 1. PAIN

lezcano commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -261,31 +261,36 @@ def exp(x):
+        if x.lower <= 0:
+            return ValueRanges.unknown()
         return ValueRanges.increasing_map(
             x, lambda y: -sympy.oo if y <= 0 else sympy.log(y)
```
With the new case, we can now simply use `smpy.log`  here?

eellison commented on test/test_value_ranges.py:
```diff
@@ -119,15 +122,45 @@ def ceil(x):
         return math.ceil(x)
 
 
+def valid_unary(fn, v):
```
we could also run the reference first and see if it throws, then skip

ezyang replied:
Still good to avoid trying to eval something silly like 2 ** 10000 which sympy will do symbolically lol

eellison commented on test/test_value_ranges.py:
```diff
@@ -153,6 +184,48 @@ def test_binary_ref(self, fn):
+            # don't attempt pow on exponents that are too large (but oo is OK)
+            if fn == "pow" and b.upper > 4 and b.upper != sympy.oo:
+                continue
+            with self.subTest(a=a, b=b):
```
`subTest` 😮

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -261,31 +261,36 @@ def exp(x):
 
     @staticmethod
     def log(x):
+        if x.lower <= 0:
```
if this is has free symbol it would throw. but we can handle that when we get here

eellison reviewed (approved):
nice !

ezyang committed:
Update on "Add exhaustive testing to ValueRanges, fix bugs"


Since I didn't want to deal with nondeterministic tests, I went the exhaustive testing route for a fixed list of constants to look at. The tests generate random ranges, propagate the range through the function, and then pick elements in the range and check that the result on the operation is in the resulting range. This caught bugs in log, sqrt and pow.

My resolution for pow was a little special, because I had trouble figuring out the correct semantics under all inputs domains. Instead, I picked two input domains (pow on two point ranges, and pow where exponent is known) and only implemented those. Everything else we give up. I think this is unlikely to affect perf.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Add exhaustive testing to ValueRanges, fix bugs"


Since I didn't want to deal with nondeterministic tests, I went the exhaustive testing route for a fixed list of constants to look at. The tests generate random ranges, propagate the range through the function, and then pick elements in the range and check that the result on the operation is in the resulting range. This caught bugs in log, sqrt and pow.

My resolution for pow was a little special, because I had trouble figuring out the correct semantics under all inputs domains. Instead, I picked two input domains (pow on two point ranges, and pow where exponent is known) and only implemented those. Everything else we give up. I think this is unlikely to affect perf.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

nunoplopes reviewed (approved):
LGTM!

ezyang committed:
Update on "Add exhaustive testing to ValueRanges, fix bugs"


Since I didn't want to deal with nondeterministic tests, I went the exhaustive testing route for a fixed list of constants to look at. The tests generate random ranges, propagate the range through the function, and then pick elements in the range and check that the result on the operation is in the resulting range. This caught bugs in log, sqrt and pow.

My resolution for pow was a little special, because I had trouble figuring out the correct semantics under all inputs domains. Instead, I picked two input domains (pow on two point ranges, and pow where exponent is known) and only implemented those. Everything else we give up. I think this is unlikely to affect perf.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

ezyang commented:
It's not intentional. You won't be able to revert this PR, please submit a forward fix PR.

yanbing-j commented:
Please review the forward fix PR https://github.com/pytorch/pytorch/pull/95688. Thanks!
